pkg/router/controller/route_secret_manager.go

```diff
@@ -118,54 +180,67 @@ func (p *RouteSecretManager) HandleRoute(eventType watch.EventType, route *route
-// DeleteFunc: Invoked when the secret is deleted. It unregisters the associated route, records the route rejection, and triggers the deletion of the route by calling the HandleRoute method with a watch.Deleted event.
+// If the validation succeeds, it updates the route's TLS certificate and key with the new secret data and calls the next plugin's HandleRoute method with a watch.Modified event, and then commits the changes by calling the next plugin's Commit() method.
+//
+// DeleteFunc: Invoked when the secret is deleted. It logs the deletion, marks the secret as deleted for this route in the deletedSecrets map,
+// records the route rejection, and triggers the deletion of the route by calling the HandleRoute method with a watch.Deleted event. NOTE: It doesn't unregister the route.
 func (p *RouteSecretManager) generateSecretHandler(route *routev1.Route) cache.ResourceEventHandlerFuncs {
 	// secret handler
 	return cache.ResourceEventHandlerFuncs{
 
-		// AddFunc is intentionally left empty (only logs the event) because this handler is generated only after ensuring the existence of the secret.
-		// By leaving this empty, we prevent unnecessary triggering for the addition of the secret again. Additionally, GetSecret() method is called
-		// immediately after registering with the secretManager, to read the secret from the cache.
 		AddFunc: func(obj interface{}) {
 			secret := obj.(*kapi.Secret)
 			log.V(4).Info("secret added for route", "namespace", route.Namespace, "secret", secret.Name, "route", route.Name)
+
+			// Secret re-creation scenario
+			// Check if the route key exists in the deletedSecrets map, indicating that the secret was previously deleted for this route.
+			// If it exists, it means the secret is being recreated. Remove the key from the map and proceed with re-registration.
+			// Otherwise, no-op (new secret creation scenario and no race condition with that flow)
+			// This helps to differentiate between a new secret creation and a re-creation of a previously deleted secret.
+			key := generateKey(route)
+			if _, deleted := p.deletedSecrets.LoadAndDelete(key); deleted {
+				log.V(4).Info("secret recreated for route", "namespace", route.Namespace, "secret", secret.Name, "route", route.Name)
+
+				// re-validate
+				// since secret re-creation will not trigger the apiserver route admission,
+				// we need to rely on the router controller for this validation.
+				if err := p.validate(route); err != nil {
+					return
+				}
+				// read the re-created secret and update TLS certificate and key
+				if err := p.populateRouteTLSFromSecret(route); err != nil {
+					return
+				}
+				// call the next plugin with watch.Modified
+				p.plugin.HandleRoute(watch.Modified, route)
+				// commit the changes
+				p.plugin.Commit()
```

Comment:
The  `RouterController` won't get an event to trigger the plugin chain in case of secret (re-)creation since the route spec won't update. 

So, it can not commit the changes in this case because that line won't be called. 
https://github.com/openshift/router/blob/72114ea78be0a032d5cdddcad694e0a352a00e93/pkg/router/controller/router_controller.go#L216

We're already doing a similar commit in case of the `UpdateFunc` logic just below it.

Reply:
That's correct.